drafts/AENS.md

```diff
@@ -0,0 +1,414 @@
+of all names and namespaces SHOULD be at most 253 characters
+long, in order to enable interoperability with DNS.
+
+
+### Registrars
+
+A registrar controls who is allowed to claim names in their namespace
+and under what circumstances they are allowed to do so.
+
+The only available registrars for this AENS version will be hard-coded
+ones, which own the `.`, `.aet` and `.test` namespaces.
+
+The `.` namespace registrar is restricted and MUST NOT allow anyone to
+claim any names in its namespace.
+
+The `.aet` namespace registrar allows users to claim any name that is
+valid according to the validation rules for labels and names if they
+are the first to claim the name and pay the appropriate fee.
+
+The `.test` namespace registrar can be, as the name suggests, used for
+test purposes
+
+
+### Commitment Scheme
+
+Claiming a name requires a commitment scheme, which enables the claimant
+to commit to a name and after the commitment has been accepted into the 
+chain, reveal the name to finish the process.
+
+A commitment should be binding, i.e. the claimant cannot change
+the value they commited to withouth changing the actual commitment
+and hiding, so that a malicious miner learns nothing about the value
+the claimant has commited until they chose to reveal that value. This
+prevents malicious miners from front running, i.e. upon seeing a 
+claim transaction, including their own claim request for the same
+name instead of the original claimant's one.
+
+
+### Hashing
+
+We are going to adopt the `NameHash` as defined by the ENS in 
+[EIP-137](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md#namehash-algorithm)
+using SHA256 instead.
+
+```
+def namehash(name):
+  if name == '':
+    return '\0' * 32
+  else:
+    label, _, remainder = name.partition('.')
+    return sha256(namehash(remainder) + sha256(label))
+```
+
+Names are generally only referred to in hashed form.
+
+### Protocol
+
+```
+                  expire
+      unclaimed <-------- revoked
+          | ^              ^^
+          | |              || expire
+          | | expire       ||
+pre-claim | |              ||  _
+          | |       revoke || | | transfer
+          v |              || | v
+     pre-claimed -------> claimed
+                  claim    | ^   
+                           | |
+                            -
+                          update
+```
+
+Note that `expire` is not an explicit message that is part
+of the protocol.
+
+
+#### Pre-claim
+
+Send a pre-claim transaction containing a hash commitment.
+
+```
+ ------------ ----
+| hash       | 32 |
+ ------------ ----
+```
+
+
+
+#### Claim
+
+```
+ ------------ ----
+| name       | 63 |
+ ------------ ----
+| fee        | 32 |
+ ------------ ----
+```
+
+Flow for a user:
+
+1. (optional) wait `n` blocks, s.t. that the block including the `pre-claim` cannot be reversed whp 
+2. send `claim` transaction to reveal name and pay the associated fee
```

Comment:
What is the relation between multiple pre-claims and the claim?
What stops a miner from posting his own pre claim and dropping the claim transaction when he sees the claim transaction? Then posting his own claim transaction?

Reply:
I guess that if an older (but live) pre claim transaction exists then the claim could go to the first pre claim this would require the malicious miner to mine all blocks till the first claim expires.

Reply:
I think a miner could only cheat by not including the original pre-claim and replacing it with a new one. This can be worked arround by including the account in the namehash.

**JUST AS IDEA**

```
def namehash(name, account):
  if name == '':
    return '\0' * 32
  else:
    label, _, remainder = name.partition('.')
    return sha256(namehash(remainder, account) + sha256(label) + account)
```
(not optimal as account is hashed for each remainder)

Reply:
@knarz whats your take on this?

Reply:
I think the simplest solution would be to forbid a `pre-claim` and the associated `claim` tx to be in the same block, or even require there to be 6 or more blocks between them. That way I can post my `pre-claim`, wait a couple of blocks and then post the claim. A malicious miner would have to mine at least two (6) blocks in a row.

I'm not sure how hashing the account together with the name would prevent a miner from just issuing their own `pre-claim` and then `claim`.

Reply:
If the `pre-claim` is just the hash of the name, wouldn't everybody be able to pre-claim the same name if the transaction is not mined yet? When hashing the account together with the name, the name itself would be hidden but in the `claim` it can be checked if this name was actually pre-claimed by the corresponding account

Reply:
Since the account (which is just a pubkey) can be recovered from a signature, the miner could still fairly easily put their own `pre-claim` into a block before a different one.
If we want to blind the value then it has to be a nonce, which would be revealed together with the name in the `claim`. This would then mean that a user would have to rely on their client software to warn them when they try to claim an already claimed name since miners would no longer be able to tell.

Anyway, I would propose that our commitment is just `Hash(Namehash + Nonce)` for now until I figure out a good way to implement auctions. The `claim` message will then have to reveal both the hash and the nonce.

Reply:
Maybe my thinking is of there, but the miner would not know what the preclaimed name is until revealed by the claim.

In case of the namehash he could pre-claim the same name without knowing what it is. (Now I understand the two blocks in a row by thinking about it.)

But with including the account in the namehash only the pre-claiming account could do the actual claim. A nonce would offer the same benefits.